contributors/design-proposals/storage/csi-migration.md

```diff
@@ -233,28 +237,71 @@ and how to redeploy a CSI Driver in a cluster.
 
 
 ## Upgrade/Downgrade, Migrate/Un-migrate
-### Kubelet Node Annotation
-When the Kubelet starts, it will check whether the feature gate is
-enabled and if so will annotate its node with `csi.attach.kubernetes.io/gce-pd`
-for example to communicate to the A/D Controller that it supports migration of
-the gce-pd to CSI. The A/D Controller will have to choose on a per-node basis
-whether to use the CSI or the in-tree plugin for attach based on 3 criterea:
-1. Feature gate
-2. Plugin Migratable (Implements MigratablePlugin interface)
-3. Node to Attach to has requisite Annotation
-
-Note: All 3 criteria must be satisfied for A/D controller to Attach/Detach with
-CSI instead of in-tree plugin. For example if a Kubelet has feature on and marks
-the annotation, but the A/D Controller does not have the feature gate flipped,
-we consider this user error and will throw some errors.
-
-This can cause a race between the A/D Controller and the Kubelet annotating, if
-a volume is attached before the Kubelet completes annotation the A/D controller
-could attach using in-tree plugin instead of CSI while the Kubelet is expecting
-a CSI Attach. The same issue exists on downgrade if the Annotation is not
-removed before a volume is attached. An additional consideration is that we
-cannot have the Kubelet downgraded to a version that does not have the
-Annotation removal code.
+
+### CSINodeInfo API
+
+`CSINodeInfoStatus` API will be modified to contain the following fields
+```
+// CSIDriverInfoStatus holds information about the status of one CSI driver installed on a node
+type CSIDriverInfoStatus struct {
+	// name is the name of the CSI driver that this object refers to.
+	// This MUST be the same name returned by the CSI GetPluginName() call for
+	// that driver.
+	Name string `json:"name"`
+
+	// available is a boolean representing whether the driver has been installed
+	// on this node or not.
```

Comment:
Should there be a `CSIDriverInfoStatus` for a driver if it is not installed/available on a node?

Reply:
`Available==false` will be for any migrate driver that is not installed, as well as any driver that has been installed then eventually uninstalled. The `CSIDriverInfoStatus`'s are never removed